src/main/java/io/supertokens/inmemorydb/queries/TOTPQueries.java

```diff
@@ -0,0 +1,197 @@
+        String QUERY = "SELECT * FROM " +
+                Config.getConfig(start).getTotpUsedCodesTable()
+                + " WHERE user_id = ? AND expiry_time_ms > ?;";
+        return execute(start, QUERY, pst -> {
+            pst.setString(1, userId);
+            pst.setLong(2, System.currentTimeMillis());
+        }, result -> {
+            List<TOTPUsedCode> codes = new ArrayList<>();
+            while (result.next()) {
+                codes.add(TOTPUsedCodeRowMapper.getInstance().map(result));
+            }
+
+            return codes.toArray(TOTPUsedCode[]::new);
+        });
+    }
+
+    public static int removeExpiredCodes(Start start)
+            throws StorageQueryException, SQLException {
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getTotpUsedCodesTable()
+                + " WHERE expiry_time_ms < ?;";
+
+        return update(start, QUERY, pst -> pst.setLong(1, System.currentTimeMillis()));
+    }
+
+    public static int removeUsedCodes(Start start, String userId, String deviceName)
+            throws StorageQueryException, SQLException {
+        // Remove codes where userId matches the given userId
+        // ONLY required for inmemorydb, as it does not support foreign key constraints.
+        String QUERY = "DELETE FROM " + Config.getConfig(start).getTotpUsedCodesTable()
+                + " WHERE user_id = ? AND device_name = ?;";
+
+        return update(start, QUERY, pst -> {
+            pst.setString(1, userId);
+            pst.setString(2, deviceName);
+        });
+    }
+
+    private static class TOTPDeviceRowMapper implements RowMapper<TOTPDevice, ResultSet> {
+        private static final TOTPDeviceRowMapper INSTANCE = new TOTPDeviceRowMapper();
+
+        private TOTPDeviceRowMapper() {
+        }
+
+        private static TOTPDeviceRowMapper getInstance() {
+            return INSTANCE;
+        }
+
+        @Override
+        public TOTPDevice map(ResultSet result) throws SQLException {
+            return new TOTPDevice(
+                    result.getString("user_id"),
+                    result.getString("device_name"),
+                    result.getString("secret_key"),
+                    result.getInt("period"),
+                    result.getInt("skew"),
+                    result.getBoolean("verified"));
+        }
+    }
+
+    private static class TOTPUsedCodeRowMapper implements RowMapper<TOTPUsedCode, ResultSet> {
+        private static final TOTPUsedCodeRowMapper INSTANCE = new TOTPUsedCodeRowMapper();
+
+        private TOTPUsedCodeRowMapper() {
+        }
+
+        private static TOTPUsedCodeRowMapper getInstance() {
+            return INSTANCE;
+        }
+
+        @Override
+        public TOTPUsedCode map(ResultSet result) throws SQLException {
+            return new TOTPUsedCode(
+                    result.getString("user_id"),
+                    result.getString("device_name"),
```

Comment:
what if this is null - will this work? - add a test for this. Add a used_code with null device_name, and get fetch used_codes for a user, and check that the returned TOTPUsedCode has deviceName as null

Reply:
TODO

Reply:
Not relevant now because we have removed `device_name` from `totp_used_codes` table.